draft-ietf-quic-http.md

```diff
@@ -469,33 +628,114 @@ The flags defined are:
+
+
+### CANCEL_REQUEST {#frame-cancel-request}
+
+The CANCEL_REQUEST frame (type=0x3) is used to request cancellation of a request
+prior to the response stream being created.  This frame can be used to cancel
+both outstanding requests initiated by a client and server push requests that
+are initiated with a PUSH_PROMISE.
+
+The frame identifies a client-initiated request by the stream ID that carried
+the request; a server push request is identified by Push ID (see
+{{frame-push-promise}}).
+
+Upon receipt of this frame the server aborts sending the response for the
+identified request.  If the server has not yet started to send the response
+stream for the identified request, it can use the receipt of a CANCEL_REQUEST
+frame to avoid opening a stream.  If a response stream or data stream has been
+opened by the server, the server SHOULD sent a QUIC RST_STREAM frame on those
+streams and cease transmission of the response.
+
+A client that cancels a request that it initiated SHOULD send a QUIC RST_STREAM
+frame to reset the associated request stream, unless that stream is already
+closed.
+
+A client sends a CANCEL_REQUEST frame on the connection control stream.  Sending
+a CANCEL_REQUEST frame on a stream other than the connection control stream MUST
+be treated as a HTTP_WRONG_STREAM error.
+
+The flags defined are:
+
+  PUSH (0x01):
+  : When set, this flag indicates that the frame identifies a push by its push
+    ID; when cleared, the frame identifies a request stream.
+
+
+~~~~~~~~~~  drawing
+    0                   1                   2                   3
+    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |                 Request Stream/Push ID (32)                   |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |                      Error Code (32)                          |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+~~~~~~~~~~
+{: #fig-cancel-request title="CANCEL_REQUEST frame payload"}
+
+The CANCEL_REQUEST frame payload has the following fields:
+
+  Request Stream/Push ID:
+  : A 32-bit stream ID or Push ID that identifies the request.  If the PUSH flag
+    is set, this refers to a Push ID included in a PUSH_PROMISE frame (see
+    {{frame-push-promise}}); if the PUSH flag is cleared, this refers to a
+    request by its stream ID.
+
+  Error Code:
+  : A 32-bit error code indicating why the response is not desired.
+
+A CANCEL_REQUEST frame always contains an 8 octet payload.  A server MUST treat
+a CANCEL_REQUEST frame payload that is other than 8 octets in length, or a
+CANCEL_REQUEST frame that identifies a stream that doesn't carry a valid request
+as an HTTP_MALFORMED_CANCEL_REQUEST error.  Note however that due to packet
+reordering, a CANCEL_REQUEST frame could arrive before the request stream that
+it references is opened.
```

Comment:
This is weird text. Isn't it always the case that if you receive an unexpected frame structure you are sad?